login/service_blackbox_test.go

```diff
@@ -0,0 +1,193 @@
+		panic("invalid test " + err.Error()) // bug
+	}
+	prms := url.Values{}
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "LoginTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeLoginContext(goaCtx, goa.New("LoginService"))
+	if err != nil {
+		panic("invalid test data " + err.Error()) // bug
+	}
+
+	err = loginService.Perform(authorizeCtx)
+
+	assert.Equal(t, 307, rw.Code)
+	assert.Contains(t, rw.Header().Get("Location"), "https://github.com/login/oauth/authorize")
+}
+
+func TestValidOAuthAuthorizationCode(t *testing.T) {
+	resource.Require(t, resource.Database)
+
+	// Current the OAuth code is generated as part of a UI workflow.
+	// Yet to figure out how to mock.
+
+}
+
+func TestValidState(t *testing.T) {
+	resource.Require(t, resource.Database)
+
+	// We do not have a test for a valid
+	// authorization code because it needs a
+	// user UI workflow. Furthermore, the code can be used
+	// only once. https://tools.ietf.org/html/rfc6749#section-4.1.2
+}
+
+func TestInvalidState(t *testing.T) {
+	resource.Require(t, resource.Database)
+
+	// Setup request context
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf("/api/login/authorize"),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	if err != nil {
+		panic("invalid test " + err.Error()) // bug
+	}
+	prms := url.Values{
+		"state": {},
+		"code":  {"invalid_code"},
+	}
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "LoginTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeLoginContext(goaCtx, goa.New("LoginService"))
+	if err != nil {
+		panic("invalid test data " + err.Error()) // bug
+	}
+
+	err = loginService.Perform(authorizeCtx)
+	assert.Equal(t, 401, rw.Code)
+}
+
+func TestInvalidOAuthAuthorizationCode(t *testing.T) {
+	resource.Require(t, resource.Database)
+
+	// We do not have a test for a valid
+	// authorization code because it needs a
+	// user UI workflow. Furthermore, the code can be used
+	// only once. https://tools.ietf.org/html/rfc6749#section-4.1.2
+
+	// Setup request context
+	rw := httptest.NewRecorder()
+	u := &url.URL{
+		Path: fmt.Sprintf("/api/login/authorize"),
+	}
+	req, err := http.NewRequest("GET", u.String(), nil)
+	req.Header.Add("referer", "localhost")
+	if err != nil {
+		panic("invalid test " + err.Error()) // bug
+	}
+	prms := url.Values{}
+	ctx := context.Background()
+	goaCtx := goa.NewContext(goa.WithAction(ctx, "LoginTest"), rw, req, prms)
+	authorizeCtx, err := app.NewAuthorizeLoginContext(goaCtx, goa.New("LoginService"))
+	if err != nil {
+		panic("invalid test data " + err.Error()) // bug
+	}
+
+	err = loginService.Perform(authorizeCtx)
+
+	assert.Equal(t, 307, rw.Code)
+
+	locationString := rw.HeaderMap["Location"][0]
+	locationUrl, err := url.Parse(locationString)
+	if err != nil {
+		t.Fatal("Redirect URL is in a wrong format ", err)
+	}
+
+	allQueryParameters := locationUrl.Query()
+	returnedState := allQueryParameters["state"][0]
```

Comment:
Good to have assert can be to check if state is present or not. In case not present, it will panic, hence.